cumulus/client/consensus/aura/src/collators/slot_based/slot_timer.rs

```diff
@@ -0,0 +1,266 @@
+// Copyright (C) Parity Technologies (UK) Ltd.
+// This file is part of Cumulus.
+// SPDX-License-Identifier: GPL-3.0-or-later WITH Classpath-exception-2.0
+
+// Cumulus is free software: you can redistribute it and/or modify
+// it under the terms of the GNU General Public License as published by
+// the Free Software Foundation, either version 3 of the License, or
+// (at your option) any later version.
+
+// Cumulus is distributed in the hope that it will be useful,
+// but WITHOUT ANY WARRANTY; without even the implied warranty of
+// MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE. See the
+// GNU General Public License for more details.
+
+// You should have received a copy of the GNU General Public License
+// along with Cumulus. If not, see <https://www.gnu.org/licenses/>.
+
+use crate::LOG_TARGET;
+use codec::Codec;
+use cumulus_primitives_aura::Slot;
+use cumulus_primitives_core::BlockT;
+use sc_client_api::UsageProvider;
+use sc_consensus_aura::SlotDuration;
+use sp_api::ProvideRuntimeApi;
+use sp_application_crypto::AppPublic;
+use sp_consensus_aura::AuraApi;
+use sp_core::Pair;
+use sp_runtime::traits::Member;
+use sp_timestamp::Timestamp;
+use std::{
+	cmp::{max, min},
+	sync::Arc,
+	time::Duration,
+};
+
+/// Lower limits of allowed block production interval.
+/// Defensive mechanism, corresponds to 12 cores at 6 second block time.
+const BLOCK_PRODUCTION_MINIMUM_INTERVAL_MS: Duration = Duration::from_millis(500);
+
+#[derive(Debug)]
+pub(crate) struct SlotInfo {
+	pub timestamp: Timestamp,
+	pub slot: Slot,
+}
+
+/// Manages block-production timings based on chain parameters and assigned cores.
+#[derive(Debug)]
+pub(crate) struct SlotTimer<Block, Client, P> {
+	/// Client that is used for runtime calls
+	client: Arc<Client>,
+	/// Offset the current time by this duration.
+	time_offset: Duration,
+	/// Last reported core count.
+	last_reported_core_num: Option<u32>,
+	/// Slot duration of the relay chain. This is used to compute how man block-production
+	/// attempts we should trigger per relay chain block.
+	relay_slot_duration: Duration,
+	_marker: std::marker::PhantomData<(Block, Box<dyn Fn(P) + Send + Sync + 'static>)>,
+}
+
+/// Compute when to try block-authoring next.
+/// The exact time point is determined by the slot duration of relay- and parachain as
+/// well as the last observed core count. If more cores are available, we attempt to author blocks
+/// for them.
+///
+/// Returns a tuple with:
+/// - `Duration`: How long to wait until the next slot.
+/// - `Timestamp`: The timestamp to pass to the inherent
+/// - `Slot`: The AURA slot used for authoring
+fn compute_next_wake_up_time(
+	para_slot_duration: SlotDuration,
+	relay_slot_duration: Duration,
+	core_count: Option<u32>,
+	time_now: Duration,
+	time_offset: Duration,
+) -> (Duration, Timestamp, Slot) {
+	let para_slots_per_relay_block =
+		(relay_slot_duration.as_millis() / para_slot_duration.as_millis() as u128) as u32;
```

Comment:
This will return 0 for a `para_slot_duration` > `relay_slot_duration`, which is not good.